src/main/java/org/truffleruby/core/FinalizationService.java

```diff
@@ -164,7 +164,7 @@ private void createFinalizationThread() {
 
     private void runFinalizer(FinalizerReference finalizerReference) {
         try {
-            while (true) {
+            while (!context.isFinalizing()) {
```

Comment:
This will break `runAllFinalizersOnExit()` in that it won't run any finalizer.

Reply:
The SafepointManager should Thread.interrupt() this Thread, so at the next `finalizerQueue::remove` it should stop.
So this isFinalizing() should only skip finalizers on this given object.